File: jsrc/machine/SxxMachine/PredicateEncoder.java
package SxxMachine;
/**
 * The <code>PredicateEncoder</code> class contains static methods for encoding predicate names.<br>
 * The predicate with <code>hoge:f/n</code> is encoded to <code>hoge.PRED_f_n</code>, where
 * <code>hoge</code> is package name,
 * <code>f</code> is predicate name, and
 * <code>n</code> is arity.<br>
 *
 * When encoding a predicate name, we apply the following rules:<br>
 *<ul>
 *<li>The alphanumeric characters
 * &ldquo;<code>a</code>&rdquo; through &ldquo;<code>z</code>&rdquo;,
 * &ldquo;<code>A</code>&rdquo; through &ldquo;<code>Z</code>&rdquo; and
 * &ldquo;<code>0</code>&rdquo; through &ldquo;<code>9</code>&rdquo; remain the same.
 *<li>The special characters &ldquo;<code>_</code>&rdquo; and &ldquo;<code>$</code>&rdquo; remain the same.
 *<li>All other characters are first converted into a list of character codes.
 * Then each character code is represented by the 5-character string &ldquo;<code>$XYZW</code>&rdquo;,
 * where <code>XYZW</code> is the four-digit hexadecimal representation of the character code.
 *</ul>
 *
 * For example,
 * a predicate with <code>hoge:(=..)/2</code> is encoded to <code>hoge.PRED_$003D$002E$002E_2</code>.
 *
 * @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 * @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 * @version 1.1
 */
public class PredicateEncoder {

    /**
     * Returns a string representation of class for
     * the predicate with the given arguments.
     * @param pkg package name
     * @param functor predicate name
     * @param arity predicate arity
     * @return a string representation of class for
     * the predicate that corresponds to <code>pkg:functor/arity</code>.
     */
    public static String encode(String pkg, String functor, int arity) {
    	char[] x = functor.toCharArray();
    	StringBuilder sb = new StringBuilder(pkg.length()+x.length*5+10);
    	sb.append(pkg);
    	sb.append(".PRED_");
    	int i = 0;
    	int s = i;
    	while (i<x.length){
    		if ( !((x[i]>='a' && x[i]<='z') || (x[i]>='A' && x[i]<='Z') ||
    				(x[i]>='0' && x[i]<='9') || x[i]=='_' || x[i]=='$')){

    			sb.append(x, s, i-s);
    			sb.append('$').append(Integer.toHexString(x[i]).substring(4, 8));
    			s = i+1;
    		}
    		i++;
    	}
    	if (s<x.length){
    		sb.append(x, s, x.length-s);
    	}
    	sb.append('_').append(arity);
    	return sb.toString();
    }

    public static String decodeFunctor(String className) {
    	int length = className.length();
    	// Remove the Java package name, if present.
    	int start = className.lastIndexOf('.');
    	start++;
    	// Trim the common PRED_ prefix.
    	if (className.regionMatches(start, "PRED_", 0, 5)){
    		start+=5;
    	}
    	// Drop the arity from the end.
    	int finish = className.lastIndexOf('_');
    	if (finish<0){
    		finish = length;
    	}

    	// replace $XXXX with character with code XXXX
    	StringBuilder sb = new StringBuilder(length);
    	int index = start;
    	int middle = index;
    	while (index<finish){
    		index = className.indexOf('$', index);

    		if (index<0){
    			index = finish;
    		} else if (index+4<finish){
    			try {
    				int c = Integer.parseInt(className.substring(index+1, index+5), 16);
    				sb.append(className.substring(middle, index));
    				switch (c) {
					case '\\':
						sb.append("\\\\");
						break;
					case '$':
						sb.append("\\$");
					default:
						sb.append((char)c);
						break;
					}
    				index = middle = index+5;
    			} catch (NumberFormatException e){
    				index++;
    			}
    		} else {
    			index++;
    		}
    	}
    	if (middle<finish){
    		sb.append(className.substring(middle, finish));
    	}
    	return sb.toString();
    }

//    public static void main(String argv[]) {
//		String p = argv[0];
//		String f = argv[1];
//		int n = (Integer.valueOf(argv[2])).intValue();
//		System.out.println(p + ":" + f + "/" + n);
//		System.out.println(PredicateEncoder.encode(p,f,n));
//    }
}

File: src/machine/SxxMachine/PredicateEncoder.cpp
using namespace std;

#include "PredicateEncoder.h"
#include "StringBuilder.h"
#include "NumberFormatException.h"

namespace SxxMachine
{

	wstring PredicateEncoder::encode(const wstring& pkg, const wstring& functor, const int& arity)
	{
		std::vector<char> x = functor.toCharArray();
		StringBuilder* sb = new StringBuilder(pkg == ""?0:pkg.length() + x.size()*5 + 10);
		sb->append("PRED_");
		int i = 0;
		int s = i;
		while(i < x.size())
		{
			if(!((x[i] >= 'a' && x[i] <= 'z') || (x[i] >= 'A' && x[i] <= 'Z') || (x[i] >= '0' && x[i] <= '9') || x[i] == '_' || x[i] == '$'))
			{
				sb->append(x, s, i - s);
				wstring hex = Long::toHexString(x[i])->toUpperCase();
				int hl = hex.length();
				if(hl > 7)
				{
					hex = hex.substr(4, 4);
				} else
				{
					while(hl < 4)
					{
						hex = "0" + hex;
						hl++;
					}
				}
				sb->append('$')->append(hex);
				s = i + 1;
			}
			i++;
		}
		if(s < x.size())
		{
			sb->append(x, s, x.size() - s);
		}
		sb->append('_')->append(arity);

		if(pkg != "")
		{
			if(!StringHelper::endsWith(pkg, "$"))
			{
				pkg += ".";
			}
		} else
		{
			pkg = "";
		}
		return pkg + sb->toString();
	}

	wstring PredicateEncoder::decodeFunctor_2(const wstring& className)
	{
		  // Remove the Java package name, if present.
		  int dot = (int)className.rfind('.');
		  if(0 < dot)
		  {
			className = className.substr(dot + 1);
		  }

		  // Trim the common PRED_ prefix.
		  if(StringHelper::startsWith(className, "PRED_"))
		  {
			className = className.substr((wstring("PRED_")).length());
		  }

		  // Drop the arity from the end.
		  int us = (int)className.rfind('_');
		  if(0 < us)
		  {
			className = className.substr(0, us);
		  }

		  Pattern* p = Pattern::compile("\\$([0-9A-F]{4})");
		  Matcher* m = p->matcher(className);
		  StringBuilder* r = new StringBuilder();
		  while(m->find())
		  {
			char c = static_cast<char>(static_cast<Integer>(m->group()));
//JAVA TO C++ CONVERTER TODO TASK: There is no native C++ equivalent to 'toString':
			m->appendReplacement(r, Character::toString(c));
		  }
		  m->appendTail(r);
		  return r->toString();
	}

	int PredicateEncoder::decodeArity(const wstring& className)
	{
	  // Drop the arity from the end.
	  int finish = (int)className.rfind('_');
	  if(finish < 0)
	  {
		 return -1;
	  }
	  if(StringHelper::endsWith(className, "_2"))
	  {
		  return 2;
	  }
	  if(StringHelper::endsWith(className, "_1"))
	  {
		  return 1;
	  }
	  if(StringHelper::endsWith(className, "_3"))
	  {
		  return 3;
	  }
	  if(StringHelper::endsWith(className, "_0"))
	  {
		  return 0;
	  }
	  if(StringHelper::endsWith(className, "_4"))
	  {
		  return 4;
	  }
	  if(StringHelper::endsWith(className, "_5"))
	  {
		  return 5;
	  }
	  if(StringHelper::endsWith(className, "_6"))
	  {
		  return 6;
	  }
	  if(StringHelper::endsWith(className, "_7"))
	  {
		  return 7;
	  }
	  if(StringHelper::endsWith(className, "_8"))
	  {
		  return 8;
	  }
	  if(StringHelper::endsWith(className, "_9"))
	  {
		  return 9;
	  }
	  return stoi(className.substr(finish) + 1);
	}

//JAVA TO C++ CONVERTER TODO TASK: Most Java annotations will not have direct C++ equivalents:
//ORIGINAL LINE: @SuppressWarnings("fallthrough") public static String decodeFunctor(String className)
	wstring PredicateEncoder::decodeFunctor(const wstring& className)
	{
		int length = className.length();
		// Remove the Java package name, if present.
		int start = (int)className.rfind('.');
		start++;
		// Trim the common PRED_ prefix.
		if(className.regionMatches(start, "PRED_", 0, 5))
		{
			start += 5;
		}
		// Drop the arity from the end.
		int finish = (int)className.rfind('_');
		if(finish < 0)
		{
			finish = length;
		}
		// replace $XXXX with character with code XXXX
		StringBuilder* sb = new StringBuilder(length);
		int index = start;
		int middle = index;
		while(index < finish)
		{
			index = (int)className.find('$', index);
			if(index < 0)
			{
				index = finish;
			} else if(index + 4 < finish)
			{
				try
				{
//JAVA TO C++ CONVERTER TODO TASK: Only single-argument parse and valueOf methods are converted:
//ORIGINAL LINE: int c = Integer.parseInt(className.substring(index+1, (index+5) - (index+1)), 16);
					int c = Integer::valueOf(className.substr(index + 1, (index + 5) - (index + 1)), 16);
					sb->append(className.substr(middle, index - middle));
					switch(c)
					{
					case '\\':
						sb->append("\\\\");
						break;
					case '$':
						sb->append("\\$");
					default:
						sb->append(static_cast<char>(c));
						break;
					}
					index = middle = index + 5;
				} catch(const NumberFormatException& e)
				{
					index++;
				}
			} else
			{
				index++;
			}
		}
		if(middle < finish)
		{
			sb->append(className.substr(middle, finish - middle));
		}
		return sb->toString();
	}

	void PredicateEncoder::main(std::vector<wstring>& argv)
	{
	wstring p = argv[0];
	wstring f = argv[1];
	int n = (stoi(argv[2])).value();
	cout << p << ":" << f << "/" << n << endl;
	cout << PredicateEncoder::encode(p,f,n) << endl;
	}
}
